### PR TITLE
fix(validator): tolerate taints, widen Trainer readiness timeout

### PR DESCRIPTION
**Pull request by mikecook**
## Summary

Make the Kubeflow Trainer/JobSet controller-manager Deployments schedulable on all-tainted node pools, and widen the Trainer controller readiness timeout to accommodate its cert-controller sidecar's normal cold-start retry latency.

## Motivation / Context

The Kubeflow Trainer/JobSet controller-manager Deployments ship with no tolerations. On a cluster where every node pool carries a taint (e.g. an arch-tainted GPU pool plus a system pool GKE reserves for its own managed components once no untainted pool remains), the controllers have nowhere to schedule and `installTrainer` times out waiting for a Deployment that can never become Ready.

Separately, on cold start the Trainer controller-manager's cert-controller sidecar provisions its webhook cert via a get-or-create against the API server; racing that against a not-yet-synced informer cache produces a `resourceVersion` conflict that the sidecar's own reconcile loop retries and self-heals from unassisted — but each retry adds latency, and on a slow cold start the cumulative delay could push first-ready past the old 2-minute budget, failing the validator's readiness wait for a controller that was already recovering on its own.

Fixes: N/A
Related: N/A

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Build/CI/tooling

## Component(s) Affected

- [ ] CLI (`cmd/aicr`, `pkg/cli`)
- [ ] API server (`cmd/aicrd`, `pkg/server`)
- [ ] Recipe engine / data (`pkg/recipe`)
- [ ] Bundlers (`pkg/bundler`, `pkg/component/*`)
- [ ] Collectors / snapshotter (`pkg/collector`, `pkg/snapshotter`)
- [x] Validator (`pkg/validator`) 
- [ ] Core libraries (`pkg/errors`, `pkg/k8s`)
- [ ] Docs/examples (`docs/`, `examples/`)
- [ ] Other: ____________

## Implementation Notes

`applyControllerTolerations` stamps a blanket tolerate-all onto the Trainer and JobSet controller-manager Deployments specifically (by name) when either has no existing tolerations; a Deployment that already declares tolerations, or any other Deployment in the manifest set, is left untouched. Scoping by name rather than by Kind alone matters here: this runs for every Deployment decoded from the installer's manifest set, and a future addition to that set must not silently inherit a blanket `{operator: Exists}` it never asked for. Built through `component.TolerationsToPodSpec`, the same converter used for Helm-values toleration overrides, so there's one canonical place that knows the toleration-to-map shape.

Also extracted the repeated `"operator"` toleration-key literal into a `keyOperator` const (used in `inference_perf_constraint.go` and `nccl_all_reduce_bw_constraint.go` too) to satisfy golangci-lint's `goconst` threshold across the package — no behavior change there.

`TrainerControllerReadyTimeout` widened from 2 to 3 minutes in `pkg/defaults/timeouts.go` — a single-constant change, independent of the toleration fix.

## Testing

```bash
make qualify
```

`make qualify` passed clean:
- `make test-coverage`: all packages pass with `-race`; `validators/performance` at 62.0% (repo-wide 84.1%, threshold 80%)
- `make lint` (golangci-lint + yamllint): clean
- `make tuning-check`: clean
- `make e2e` (chainsaw, `--no-cluster`): 24/24 passed, 0 failed, 0 skipped
- `make scan` (grype): no new vulnerabilities introduced by this change
- `make license-check`: clean
- `make api-diff`: no incompatible SDK facade / transparent-alias changes since v0.20.0

Added `trainer_lifecycle_test.go` coverage for `applyControllerTolerations` (existing tolerations preserved, blanket toleration applied when absent, unrelated Deployments/Kinds untouched).

## Risk Assessment

- [x] **Low** — Isolated change, well-tested, easy to revert
- [ ] **Medium** — Touches multiple components or has broader impact
- [ ] **High** — Breaking change, affects critical paths, or complex rollout

**Rollout notes:** N/A — internal validator behavior only. No recipe/API/CLI surface change; the Trainer/JobSet controller Deployments now tolerate all taints and get one extra minute to become ready, nothing else changes.

## Checklist

- [x] Tests pass locally (`make test` with `-race`)
- [x] Linter passes (`make lint`)
- [x] I did not skip/disable tests to make CI green
- [x] I added/updated tests for new functionality
- [ ] I updated docs if user-facing behavior changed 
- [x] Changes follow existing patterns in the codebase
- [x] Commits are cryptographically signed (`git commit -S`) — [GPG signing info](https://docs.github.com/en/authentication/managing-commit-signature-verification)

### CODE REVIEW

**Review by njhensley — COMMENTED** [review]
## Multi-persona review — tolerate taints + widen Trainer readiness timeout

**Method:** 4 independent persona reviewers (Correctness · Domain/K8s-Architecture · Test-coverage · Operability/CI-DX), every finding then adversarially re-derived from the resolved code by a senior meta-reviewer. Anchored to head `793f445d`.

**Tier legend:** 🔴 Blocker · 🟠 Major · 🟡 Minor · 🔵 Nitpick

### Overall assessment

Small, well-scoped, and genuinely clean. It makes the self-install Trainer/JobSet controller-manager Deployments schedulable on all-tainted clusters (blanket `{operator: Exists}`) and gives the Trainer controller one extra minute to become ready. **Nothing survived adversarial re-derivation at 🟠 or above.** The two scariest-looking candidates both dissolved under the code:

- The shared-package-var **aliasing** of `controllerTolerateAll` is real but inert — both apply paths `obj.DeepCopy()` before Create/Update and nothing mutates tolerations in place.
- The **name-vs-label asymmetry** (toleration matches by name, readiness lookup matches by label) is correct *by construction*: `applyControllerTolerations` mutates the pre-apply kustomize manifest whose name is guaranteed by the overlay, while the readiness lookup queries the *live* cluster where a Helm release could rename it.

Routing the toleration shape through `component.TolerationsToPodSpec` is good hygiene, and the test quality is high — assertions verify content (not just length) and correctly distinguish "field absent" from "field empty."

**Recommendation: ✅ Approve with comments.** One Minor worth a look; the rest are optional polish. Single most valuable follow-up: make the name-based match observable (warn-log or decode-time assertion) so a future upstream rename fails loudly rather than as a bare readiness timeout.

### Confirmed non-issues (examined, refuted or by-design)

- **Aliasing → no live bug** — DeepCopy at both apply sites; nothing mutates tolerations in place.
- **Name-vs-label asymmetry** — correct by construction (pre-apply manifest vs. live cluster).
- **3m timeout** — well within budget; `NCCLTrainJobTimeout=30m` dominates, each wait has its own `ctx.WithTimeout`, and `timeouts_test` bounds it to [1m, 5m].
- **Blanket tolerate-all with no nodeSelector** — acceptable by design; on an all-tainted cluster the controllers must land *somewhere* and a hard nodeSelector is infeasible.
- **Fix only covers the self-install path, not recipe-declared Helm trainers** — correct ownership separation; mutating a recipe-delivered Deployment would be the actual defect.
- **Idempotency/re-run** — guard reads the fresh render, not the live object; re-install re-stamps and `Update` converges.
- **Docs** — internal-only ephemeral fixture; no user-facing surface, no docs update required.

### Summary

| 🔴 Blocker | 🟠 Major | 🟡 Minor | 🔵 Nitpick |
|:---:|:---:|:---:|:---:|
| 0 | 0 | 1 | 9 |

*Reviewed with a multi-persona panel + adversarial meta-review. Inline comments follow.*

File: validators/performance/trainer_lifecycle.go [hunk 1]
```diff
@@ -159,6 +166,57 @@ const (
...
+// applyControllerTolerations stamps controllerTolerateAll onto the Trainer and
+// JobSet controller-manager Deployments' pod template, unless one already
+// declares tolerations. Scoped to those two names so an unrelated Deployment
+// in the manifest set never gets a blanket toleration it didn't ask for.
+func applyControllerTolerations(obj *unstructured.Unstructured) error {
+	if obj.GroupVersionKind().Kind != "Deployment" {
+		return nil
+	}
+	switch obj.GetName() {
+	case trainerControllerDeployment, jobSetControllerDeployment:
```
Comment (njhensley) [inline validators/performance/trainer_lifecycle.go:216]: **🟡 Minor — Name-based match silently no-ops if a future archive bump renames a controller**

The two controller Deployments are matched by exact name in this `switch`. The Kubeflow Trainer archive is downloaded at runtime, so a future pinned-version bump (or the separately-managed JobSet overlay variant) that renames the emitted Deployment makes both cases miss — the function returns `nil` with no log or error, and that controller reverts to un-schedulable on all-tainted clusters. It resurfaces as a bare (now 3-minute) readiness timeout, precisely the failure the label-based readiness lookup was built to avoid.

**Blast radius:** Self-install path; the JobSet/Trainer controller silently loses its toleration on a future upstream rename and the benchmark fails at readiness wait rather than with a clear signal.

**Fix:** Emit a warn-log when a controller-shaped Deployment isn't matched, or add a decode-time assertion that both expected controller names appeared among the decoded objects.

File: pkg/defaults/timeouts.go [hunk 2]
```diff
@@ -667,7 +667,7 @@ const (
 
 	// TrainerControllerReadyTimeout is the time to wait for the Kubeflow Trainer
 	// controller-manager Deployment to have at least one ready replica after installation.
-	TrainerControllerReadyTimeout = 2 * time.Minute
+	TrainerControllerReadyTimeout = 3 * time.Minute
```
Comment (njhensley) [inline pkg/defaults/timeouts.go:677]: **🔵 Nitpick — 2m→3m rationale isn't captured in-code**

The value changed 2m→3m but the doc comment still describes only "at least one ready replica after installation." The cert-controller cold-start / retry reasoning lives only in the PR body and is lost on squash-merge.

**Fix:** Fold a one-line why into the comment so the next engineer doesn't tune it back down and reintroduce the flake.

File: validators/performance/consts.go [hunk 3]
```diff
@@ -21,6 +21,7 @@ const (
 	versionV1alpha1          = "v1alpha1"
 	versionV1beta1           = "v1beta1"
 	keyName                  = "name"
+	keyOperator              = "operator"
```
Comment (njhensley) [inline validators/performance/consts.go:24]: **🔵 Nitpick — keyOperator const wasn't actually required by goconst**

The PR justifies this const as needed "to satisfy golangci-lint's goconst." But `.golangci.yaml` sets `goconst: {min-occurrences: 3, ignore-tests: true}`, and pre-PR there were exactly 2 non-test occurrences of `"operator"` (the two source files this PR edits) — goconst was already passing.

**Fix:** None required — the extraction is a fine readability improvement; just noting the stated CI justification is inaccurate.

File: validators/performance/trainer_lifecycle.go [hunk 4]
```diff
@@ -159,6 +166,57 @@ const (
 	jobSetPromotedImageRepo = "registry.k8s.io/jobset/jobset"
 )
 
+// controllerTolerateAll lets a Trainer/JobSet controller-manager Deployment
+// schedule on any node pool, regardless of taints. Built through
+// component.TolerationsToPodSpec, the same converter used for Helm-values
+// toleration overrides, so there is one canonical place that knows the
+// toleration-to-map shape.
+var controllerTolerateAll = tolerationsToAnySlice(
```
Comment (njhensley) [inline validators/performance/trainer_lifecycle.go:174]: **🔵 Nitpick — controllerTolerateAll is aliased by reference into both Deployments (latent)**

`controllerTolerateAll` is a package-level `[]any` holding one shared `map[string]any`, assigned by reference to `podSpec["tolerations"]` for both the Trainer and JobSet Deployments. No live bug today — both apply paths `obj.DeepCopy()` before Create/Update and nothing mutates the tolerations in place — but a future in-place edit of one controller's tolerations would silently corrupt the other and the process-wide global.

**Fix:** A per-call copy inside applyControllerTolerations is cheap hygiene, and a two-Deployment isolation test would pin it.

File: validators/performance/trainer_lifecycle.go [hunk 5]
```diff
@@ -159,6 +166,57 @@ const (
...
+	switch obj.GetName() {
+	case trainerControllerDeployment, jobSetControllerDeployment:
+	default:
+		return nil
+	}
+
+	if existing, found, err := unstructured.NestedSlice(obj.Object, "spec", "template", "spec", "tolerations"); err != nil {
+		return aicrErrors.Wrap(aicrErrors.ErrCodeInternal,
+			fmt.Sprintf("failed to read tolerations from Deployment %q", obj.GetName()), err)
+	} else if found && len(existing) > 0 {
```
Comment (njhensley) [inline validators/performance/trainer_lifecycle.go:224]: **🔵 Nitpick — "already tolerated → skip" branch returns silently**

The apply branch logs `slog.Info("Applying blanket toleration…")`, but this `found && len(existing) > 0 { return nil }` skip branch returns with no log. An operator debugging why a controller is still Pending on a tainted pool gets no signal that the blanket toleration was deliberately withheld because the manifest already had its own.

**Fix:** Add a `slog.Debug("…already declares tolerations; leaving untouched", "name", obj.GetName())` before the early return.

File: validators/performance/trainer_lifecycle.go [hunk 6]
```diff
@@ -159,6 +166,57 @@ const (
...
+		return nil
+	}
+
+	podSpec, found := nestedMap(obj.Object, "spec", "template", "spec")
+	if !found {
+		return aicrErrors.New(aicrErrors.ErrCodeInternal,
+			fmt.Sprintf("pod spec not found in Deployment %q", obj.GetName()))
+	}
+	podSpec["tolerations"] = controllerTolerateAll
+	slog.Info("Applying blanket toleration to controller Deployment", "name", obj.GetName())
```
Comment (njhensley) [inline validators/performance/trainer_lifecycle.go]: **🔵 Nitpick — Applied-toleration log omits namespace**

This log is name-only; the surrounding Deployment-lifecycle logs in this file include `"namespace"`. The overlay sets namespace, so it's available at decode time.

**Fix:** Add `"namespace", obj.GetNamespace()` for symmetry.

File: validators/performance/trainer_lifecycle.go [hunk 7]
```diff
@@ -159,6 +166,57 @@ const (
...
+	}
+	return result
+}
+
+// applyControllerTolerations stamps controllerTolerateAll onto the Trainer and
+// JobSet controller-manager Deployments' pod template, unless one already
+// declares tolerations. Scoped to those two names so an unrelated Deployment
+// in the manifest set never gets a blanket toleration it didn't ask for.
+func applyControllerTolerations(obj *unstructured.Unstructured) error {
+	if obj.GroupVersionKind().Kind != "Deployment" {
```
Comment (njhensley) [inline validators/performance/trainer_lifecycle.go]: **🔵 Nitpick — Kind check matches string only, not apps group**

`obj.GroupVersionKind().Kind != "Deployment"` gates on the Kind string alone; a `Deployment` kind in a non-`apps` group whose name collided with a controller name would be stamped. Harmless for the kustomize overlay this runs over (only the two standard apps/v1 Deployments), so purely defensive.

**Fix:** Optionally also assert `obj.GroupVersionKind().Group == "apps"`.

File: validators/performance/trainer_lifecycle.go (hunk 5 above)
Comment (njhensley) [inline validators/performance/trainer_lifecycle.go:224]: **🔵 Nitpick — present-but-empty tolerations boundary untested**

The guard `found && len(existing) > 0` means a Deployment shipping `tolerations: []` falls through and gets stamped (the desired behavior). The test covers only nil and non-empty.

**Fix:** Add a `deploymentFixture(trainerControllerDeployment, []any{})` case pinning this boundary against a future refactor that flips the guard to `found` alone.

File: validators/performance/trainer_lifecycle.go [hunk 8]
```diff
@@ -807,6 +865,9 @@ func decodeTrainerObjects(resources []*resource.Resource) ([]*unstructured.Unstr
 		if obj.GroupVersionKind().Kind == "" {
 			continue
 		}
+		if tolErr := applyControllerTolerations(obj); tolErr != nil {
```
Comment (njhensley) [inline validators/performance/trainer_lifecycle.go:888]: **🔵 Nitpick — decodeTrainerObjects has no end-to-end test**

`applyControllerTolerations` is unit-tested directly (both error paths included), but its only call site — decodeTrainerObjects — has no test. The wiring (toleration actually lands on returned objects, decode-error propagation, the `Kind==""` skip ordering just above this call) is unverified.

**Fix:** Feed decodeTrainerObjects a small fixture with a controller Deployment + an unrelated object and assert the toleration lands only on the controller (and that a malformed resource yields a non-nil error).

File: validators/performance/nccl_all_reduce_bw_constraint.go [hunk 9]
```diff
@@ -1451,7 +1451,7 @@ func applyNCCLWorkerScheduling(obj *unstructured.Unstructured, nodeSelector map[
 			tolList := make([]any, 0, len(tolerations))
 			for _, t := range tolerations {
 				tolMap := map[string]any{
-					"operator": string(t.Operator),
+					keyOperator: string(t.Operator),
```
Comment (njhensley) [inline validators/performance/nccl_all_reduce_bw_constraint.go:1454]: **🔵 Nitpick — keyOperator output key unpinned by tests**

The `tolerationsToUnstructured` / `applyNCCLWorkerScheduling` tests assert `key/value/effect` but never read `tol["operator"]`, so the const-keyed output key is compiler-checked but not test-pinned.

**Fix:** Add `tol["operator"]` to the existing key/value/effect assertions to harden the refactor.

File: validators/performance/trainer_lifecycle_test.go [hunk 10]
```diff
@@ -76,3 +81,281 @@ func TestRewriteJobSetStagingImage_PreservesTag(t *testing.T) {
...
+	}}
+}
+
+// TestApplyControllerTolerations covers both controller names, the two
+// mutation-failure paths, and that an unrelated Deployment is left untouched.
+// TestApplyControllerTolerations_Isolation pins that the two controllers do
+// not share a live toleration slice: mutating one Deployment's stamped
+// tolerations in place must not affect the other, or the shared
+// controllerTolerateAll package-level global.
+func TestApplyControllerTolerations_Isolation(t *testing.T) {
```
Comment (njhensley) [inline validators/performance/trainer_lifecycle_test.go:114]: **🟡 Minor — Isolation test is a false green — NestedSlice deep-copies, so it never exercises aliasing**

`unstructured.NestedSlice` returns `runtime.DeepCopyJSONValue(val)`, so `trainerTols[0]` is a deep copy — mutating `trainerTol["key"]` never touches the live Deployment or the `controllerTolerateAll` global.

**Blast radius:** Reverting `cloneControllerTolerateAll` to the aliasing behavior leaves `TestApplyControllerTolerations_Isolation` passing (verified empirically), so it does not catch the regression it was added (prior P4) to pin. The production clone itself is correct — this is only the guard.

**Fix:** Read the live map directly (`NestedFieldNoCopy`, or index into `obj.Object[...]`) before mutating, or assert the two stamped toleration maps are distinct instances (mutate one via the live pod-spec map and confirm the other Deployment + the global are unchanged).

File: validators/performance/trainer_lifecycle_test.go [hunk 11]
```diff
@@ -76,3 +81,281 @@ func TestRewriteJobSetStagingImage_PreservesTag(t *testing.T) {
...
+		"metadata":   map[string]any{"name": name},
+		"spec": map[string]any{
+			"template": map[string]any{
+				"spec": podSpec,
+			},
+		},
+	}}
+}
+
+// TestApplyControllerTolerations covers both controller names, the two
```
Comment (njhensley) [inline validators/performance/trainer_lifecycle_test.go:108]: **🔵 Nitpick — Stranded doc comment: opens describing TestApplyControllerTolerations but precedes the _Isolation func**

The comment block's first two sentences describe `TestApplyControllerTolerations`, but the block sits directly above `func TestApplyControllerTolerations_Isolation` (godoc attributes it there), while the table-driven `TestApplyControllerTolerations` at :139 has no doc of its own.

**Fix:** Move the first two sentences down to immediately above `func TestApplyControllerTolerations`.

File: validators/performance/trainer_lifecycle.go [hunk 12]
```diff
@@ -807,8 +885,27 @@ func decodeTrainerObjects(resources []*resource.Resource) ([]*unstructured.Unstr
...
+			return nil, tolErr
+		}
+		if obj.GroupVersionKind().Kind == "Deployment" {
+			seenControllers[obj.GetName()] = true
```
Comment (njhensley) [inline validators/performance/trainer_lifecycle.go:892]: **🔵 Nitpick — seenControllers warn-loop key not apps-group scoped (inert asymmetry)**

`seenControllers[obj.GetName()]` records any `Kind=="Deployment"` regardless of group, whereas `applyControllerTolerations` now also gates on `gvk.Group=="apps"`. A non-apps Deployment-kind named like a controller would mark the name "seen" and suppress the warn while not receiving the toleration — the exact silent-miss the warn guards.

**Blast radius:** Unreachable in the fixed upstream kustomize set; purely defensive.

**Fix:** Optionally gate the `seenControllers` write on `gvk.Group=="apps"` too, to match the apply gate.

**Review by njhensley — APPROVED** [review]
## Re-review — tolerate taints + widen Trainer readiness timeout

**Mode:** re-review (delta). My prior review (1 🟡 + 9 🔵, "Approve with comments") stood against `07dcd0f2`; **11 commits landed since**, each a direct response to a prior finding. Re-anchored to head `4ae436aa`.

**Method:** every prior finding re-verified against the resolved code, plus one adversarial pass for net-new defects. The crux net-new finding (isolation-test false-green) was proven empirically with a throwaway reproduction, not just argued.

**Tier legend:** 🔴 Blocker · 🟠 Major · 🟡 Minor · 🔵 Nitpick

### Overall assessment

A model fix round. All 10 prior findings are resolved, and the two that mattered most — the name-based silent-no-op (🟡) and the shared-var aliasing (🔵) — got the treatment they warranted: a decode-time warn loop, and a real per-call clone. Tests pass, and the new warn loop does **not** fire spuriously: the kustomize overlay emits *both* controller Deployments in the one manifest set, so both names are always seen on a healthy run.

One net-new snag: the isolation **test** added to lock in the aliasing fix reads through `unstructured.NestedSlice`, which deep-copies — so its mutation never reaches the live object or the global. Reverting the clone leaves the test green (proven empirically): a guard that doesn't guard. The *production* clone is correct, so there's no live bug.

**Approving with comments.** Nothing at 🟠+. The one follow-up worth doing is tightening that isolation test (N1).

### Prior-feedback status

| # | Prior finding | Disposition | Evidence |
|---|---|---|---|
| P1 🟡 | Name match silent no-op on rename | ✔️ Addressed | decode-time warn loop; both controllers in one manifest set → no spurious fire |
| P2 🔵 | 2m→3m rationale not in-code | ✔️ Addressed | expanded comment |
| P3 🔵 | keyOperator const not required by goconst | ⊘ No code change needed | finding said "none required"; const is fine. The PR body still asserts the goconst justification — inaccurate but harmless |
| P4 🔵 | controllerTolerateAll aliased | ✔️ Prod fixed / ◐ test | clone is correct; the isolation test pinning it is ineffective → **N1 inline** |
| P5 🔵 | already-tolerated skip silent | ✔️ Addressed | `slog.Debug` added |
| P6 🔵 | applied-toleration log omits namespace | ✔️ Addressed | namespace added |
| P7 🔵 | Kind check not apps-group gated | ✔️ Addressed | `gvk.Group=="apps"` + test |
| P8 🔵 | present-but-empty tolerations untested | ✔️ Addressed | boundary case |
| P9 🔵 | decodeTrainerObjects no e2e test | ✔️ Addressed | `TestDecodeTrainerObjects(+_PropagatesTolerationError)` |
| P10 🔵 | keyOperator output key untested | ✔️ Addressed | asserts `operator=Equal` |

### Confirmed non-issues (net-new pass)

- **`controllerTolerateAll` package-var init** — `component.TolerationsToPodSpec` is a pure in-memory transform; `[{Operator:Exists}]` → `[{"operator":"Exists"}]`, non-empty. Safe at init.
- **`cloneControllerTolerateAll` shallow map copy** — sufficient; every `TolerationsToPodSpec` value is a scalar (no nested maps/slices to alias).
- **Warn-loop spurious fire** — refuted; both controller Deployments ship in the single kustomize manifest set.
- **Table-test dispositions** — `wantTolerations==nil` ⇒ field-absent vs. `wantErr` for missing-pod-spec / malformed-tolerations are correctly distinguished; length checked before per-key compare.

### Summary

| 🔴 Blocker | 🟠 Major | 🟡 Minor | 🔵 Nitpick |
|:---:|:---:|:---:|:---:|
| 0 | 0 | 1 | 2 |

All prior feedback addressed. Inline comments follow (N1–N3).
